test/unit/controllers/declaration/WarehouseIdentificationControllerSpec.scala

```diff
@@ -0,0 +1,110 @@
+import forms.Choice.AllowedChoiceValues.SupplementaryDec
+import forms.declaration.TransportCodes.Maritime
+import helpers.views.declaration.WarehouseIdentificationMessages
+import org.mockito.ArgumentMatchers.any
+import org.mockito.Mockito
+import org.mockito.Mockito.{times, verify, when}
+import org.scalatest.BeforeAndAfterEach
+import play.api.libs.json.Json
+import play.api.test.Helpers._
+import play.twirl.api.HtmlFormat
+import unit.base.ControllerSpec
+import views.html.declaration.warehouse_identification
+
+class WarehouseIdentificationControllerSpec
+    extends ControllerSpec with BeforeAndAfterEach with WarehouseIdentificationMessages {
+
+  val warehouseIdentificationTemplate: warehouse_identification = mock[warehouse_identification]
+
+  val controller = new WarehouseIdentificationController(
+    authenticate = mockAuthAction,
+    journeyType = mockJourneyAction,
+    exportsCacheService = mockExportsCacheService,
+    mcc = stubMessagesControllerComponents(),
+    warehouseIdentificationPage = warehouseIdentificationTemplate
+  )
+
+  val exampleCustomsOfficeIdentifier = "A1B2C3D4"
+  val exampleWarehauseIdentificationType = "R"
+  val exampleWarehauseIdentificationNumber = "SecretStash"
+  val exampleTransportMode = Maritime
+
+  val cacheModel = aCacheModel(
+    withChoice(SupplementaryDec),
+    withWarehouseIdentification(
+      Some(exampleCustomsOfficeIdentifier),
+      Some(exampleWarehauseIdentificationType),
+      Some(exampleWarehauseIdentificationNumber),
+      Some(exampleTransportMode)
+    )
+  )
+
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
+    authorizedUser()
+    withNewCaching(cacheModel)
+    when(warehouseIdentificationTemplate.apply(any())(any(), any())).thenReturn(HtmlFormat.empty)
+  }
+
+  override protected def afterEach(): Unit = {
+    Mockito.reset(warehouseIdentificationTemplate)
+    super.afterEach()
+  }
+
+  "WerehouseIdentificationController on GET request" should {
+    "return 200 OK" in {
+      val response = controller.displayForm().apply(getRequest())
+      status(response) must be(OK)
+    }
+
+    "read item from cache and display it" in {
+      val result = controller.displayForm().apply(getRequest())
+      await(result)
+      verify(mockExportsCacheService, times(2)).get(any())
+      verify(warehouseIdentificationTemplate).apply(any())(any(), any())
+    }
+  }
+  "Warehouse Identification Controller on POST" should {
+    "update cache after successful bind" in {
+      val body = Json.obj(
+        "supervisingCustomsOffice" -> exampleCustomsOfficeIdentifier,
+        "identificationType" -> exampleWarehauseIdentificationType,
+        "identificationNumber" -> exampleWarehauseIdentificationNumber,
+        "inlandModeOfTransportCode" -> exampleTransportMode
+      )
+
+      val result = controller.saveWarehouse().apply(postRequest(body))
+
+      await(result)
+
+      verify(mockExportsCacheService).update(any(), any())
```

Comment:
We should verify what we call update with.

There is some helper methods e.g. `theCacheModelUpdated` which uses argument captors allowing you tot assert on the content

Reply:
Maybe we should also start checking `id` of model we are going to update ?

Reply:
If we were being thorough then yes, but because it's generated it's a bit awkward. I guess we could mock the `journeySessionId` or whatever generates it